java-checks/src/main/java/org/sonar/java/checks/VolatileVariablesIncrementDecrementCheck.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+
+@Rule(key = "S3078")
+public class VolatileVariablesIncrementDecrementCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE_TEMPLATE = "Use an \"%s\" for this field; its %s are atomic.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.PREFIX_INCREMENT, Tree.Kind.POSTFIX_INCREMENT, Tree.Kind.PREFIX_DECREMENT, Tree.Kind.POSTFIX_DECREMENT, Tree.Kind.LOGICAL_COMPLEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    UnaryExpressionTree unaryExpression = (UnaryExpressionTree) tree;
+    ExpressionTree expression = ExpressionUtils.skipParentheses(unaryExpression.expression());
+    IdentifierTree identifier = getVariableIdentifier(expression);
+
+    if (identifier == null || !identifier.symbol().isVolatile()) {
```

Comment:
because you use semantic here, you should be sure that we have access to it, and start the `visitNode()` method with semantic check:
```
if (!hasSemantic()) {
  return;
}
```